bigtable-client-core-parent/bigtable-client-core/src/test/java/com/google/cloud/bigtable/grpc/async/TestBulkMutationGCJClient.java

```diff
@@ -0,0 +1,184 @@
+    future.setException(new RuntimeException(message));
+    ApiFuture<Void> result = bulkMutationClient.add(rowMutation);
+    Assert.assertTrue(result.isDone());
+    result.get();
+    expect.expectMessage(message);
+    verify(callable).futureCall(rowMutation);
+  }
+
+  @Test
+  public void testCallableTooFewStatuses() {
+    RowMutation rowMutationOther = RowMutation.create("table", "key");
+    SettableApiFuture<Void> futureOther = SettableApiFuture.create();
+    when(callable.futureCall(rowMutation)).thenReturn(future);
+    when(callable.futureCall(rowMutationOther)).thenReturn(futureOther);
+
+    ApiFuture<Void> result = bulkMutationClient.add(rowMutation);
+    ApiFuture<Void> resultOther = bulkMutationClient.add(rowMutationOther);
+    // Only resolving one response
+    future.set(null);
+
+    Assert.assertTrue(result.isDone());
+    Assert.assertFalse(resultOther.isDone());
+
+    Assert.assertFalse(bulkMutationClient.isFlushed());
+    futureOther.set(null);
+    Assert.assertTrue(bulkMutationClient.isFlushed());
+
+    verify(callable).futureCall(rowMutation);
+    verify(callable).futureCall(rowMutationOther);
+  }
+
+  @Test
+  public void testConcurrentBatches() throws Exception {
+    final List<ApiFuture<Void>> futures =
+        Collections.synchronizedList(new ArrayList<ApiFuture<Void>>());
+    final int batchCount = 10;
+    final int concurrentBulkMutationCount = 50;
+
+    Runnable r = new Runnable() {
+      @Override
+      public void run() {
+        IBulkMutation bulkMutation = new BulkMutationGCJClient(batcher);
+        for (int i = 0; i < batchCount * 10; i++) {
+          futures.add(bulkMutation.add(RowMutation.create("fake-table", "fake-key")));
+        }
+        bulkMutation.sendUnsent();
+      }
+    };
+    ExecutorService pool = Executors.newFixedThreadPool(100);
+    for (int i = 0; i < concurrentBulkMutationCount; i++) {
+      pool.execute(r);
+    }
+    pool.shutdown();
+    pool.awaitTermination(100, TimeUnit.SECONDS);
+
+    for (ApiFuture<Void> future : futures) {
+      Assert.assertTrue(future.isDone());
+    }
+    pool.shutdownNow();
+  }
+
+  class BatcherUnaryCallable extends UnaryCallable<RowMutation, Void> {
+    @Override
+    public ApiFuture<Void> futureCall(RowMutation rowMutation, ApiCallContext apiCallContext) {
+      return ApiFutures.immediateFuture(null);
+    }
+  }
+
+  @Test
+  public void testIsFlush() {
+    when(callable.futureCall(rowMutation)).thenReturn(future);
+    bulkMutationClient.add(rowMutation);
+    Assert.assertFalse("BulkMutation should have one pending element",
+        bulkMutationClient.isFlushed());
+    future.set(null);
+    Assert.assertTrue("BulkMutation should not have any pending element",
+        bulkMutationClient.isFlushed());
+    verify(callable).futureCall(rowMutation);
+  }
+
+  @Test
+  public void testFlush() {
+    final ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1);
```

Comment:
Why do you need this scheduler? Can't this just be done in line?  If you do need it, please make sure to close it in a finally block.

Reply:
We need the scheduler to test the flush operation is blocking until every response is resolved.

Have added scheduler into a `finally` block.